src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/BinaryFormatterUnsafeDeserializeAnalyzer.cs

```diff
@@ -39,71 +38,81 @@ public override void Initialize(AnalysisContext context)
+                // bail out
                 return;
             }
 
-            // If the accessed identifier resolves to a type symbol other than BinaryFormatter, then bail out
-            // since it means the user is calling some other similarly named API.
+            // at this point, we have confirmed that the code is invoking a method named "UnsafeDeserialize"
+            // next we want to see if this method belongs to System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
+            var accessedIdentifier = expression.GetAccessedIdentifier();
+
             var accessedSymbol = context.SemanticModel.GetSymbolInfo(accessedIdentifier).Symbol;
-            if (accessedSymbol is ILocalSymbol localSymbol)
+            if (accessedSymbol != null
+                && !IsSymbolAVariableOfTheTypeBinaryFormatter(accessedSymbol)
+                && !IsSymbolAConstructorInstanceOfBinaryFormatter(accessedSymbol))
             {
-                // refactored from $"{localSymbol.Type.ContainingNamespace}.{localSymbol.Type.Name}" for performance reasons
-                // avoiding the memory pressure from creating an interpolated string due to the volume of times a syntax analyzer is executed
-                if (!TargetTypeSymbolNamespace.Equals(localSymbol.Type.ContainingNamespace.ToString(), StringComparison.Ordinal)
-                    || !TargetTypeSymbolName.Equals(localSymbol.Type.Name, StringComparison.Ordinal))
-                {
-                    return;
-                }
+                // we found the type that owns this method
+                // and we proved it is not "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter"
+                return;
             }
-            else if (accessedSymbol is INamedTypeSymbol symbol)
+
+            var diagnostic = Diagnostic.Create(Rule, expression.GetLocation());
+            context.ReportDiagnostic(diagnostic);
+        }
+
+        /// <summary>
+        /// Checks to see if the symbol is a reference to BinaryFormatter.
+        /// </summary>
+        /// <param name="theSymbol">any symbol.</param>
+        /// <returns>True when (e.g. new BinaryFormatter().UnsafeDeserialize and <paramref name="theSymbol"/> is the symbol representing BinaryFormatter).</returns>
+        private static bool IsSymbolAConstructorInstanceOfBinaryFormatter(ISymbol theSymbol)
+        {
+            if (theSymbol is null)
             {
-                var symbolText = symbol.ToDisplayString(NullableFlowState.NotNull);
-                if (!symbolText.StartsWith(TargetTypeSymbolNamespace, StringComparison.Ordinal)
-                    || !symbolText.EndsWith(TargetTypeSymbolName, StringComparison.Ordinal))
-                {
-                    return;
-                }
+                return false;
             }
-            else if (accessedSymbol != null)
+
+            return theSymbol.ContainingType.ToString().Equals(QualifiedTargetSymbolName, StringComparison.Ordinal);
```

Comment:
Once my PR for the HttpContext refactor gets in, I've got a `NameMatcher` helper that will do this in a better way (ie checks each part rather than calling tostring)

Reply:
i saw you remembered VB case insensitive, so this isn't actually correct as-is